### PR TITLE
For PHP 8

### PR DESCRIPTION
**Pull request by remicollet**
First commit fix all build warnings

Second fix compatibility with PHP 8 adding arginfo for all functions

### CODE REVIEW

**Comment by remicollet** [comment]
Reflection output

```
Extension [ <persistent> extension #91 PC/SC version 0.4.0 ] {

  - Constants [74] {
    Constant [ int SCARD_E_BAD_SEEK ] { 2148532265 }
    Constant [ int SCARD_E_CANCELLED ] { 2148532226 }
    Constant [ int SCARD_E_CANT_DISPOSE ] { 2148532238 }
    Constant [ int SCARD_E_CARD_UNSUPPORTED ] { 2148532252 }
    Constant [ int SCARD_E_CERTIFICATE_UNAVAILABLE ] { 2148532269 }
    Constant [ int SCARD_E_COMM_DATA_LOST ] { 2148532271 }
    Constant [ int SCARD_E_DIR_NOT_FOUND ] { 2148532259 }
    Constant [ int SCARD_E_DUPLICATE_READER ] { 2148532251 }
    Constant [ int SCARD_E_FILE_NOT_FOUND ] { 2148532260 }
    Constant [ int SCARD_E_ICC_CREATEORDER ] { 2148532257 }
    Constant [ int SCARD_E_ICC_INSTALLATION ] { 2148532256 }
    Constant [ int SCARD_E_INSUFFICIENT_BUFFER ] { 2148532232 }
    Constant [ int SCARD_E_INVALID_ATR ] { 2148532245 }
    Constant [ int SCARD_E_INVALID_CHV ] { 2148532266 }
    Constant [ int SCARD_E_INVALID_HANDLE ] { 2148532227 }
    Constant [ int SCARD_E_INVALID_PARAMETER ] { 2148532228 }
    Constant [ int SCARD_E_INVALID_TARGET ] { 2148532229 }
    Constant [ int SCARD_E_INVALID_VALUE ] { 2148532241 }
    Constant [ int SCARD_E_NO_ACCESS ] { 2148532263 }
    Constant [ int SCARD_E_NO_DIR ] { 2148532261 }
    Constant [ int SCARD_E_NO_FILE ] { 2148532262 }
    Constant [ int SCARD_E_NO_KEY_CONTAINER ] { 2148532272 }
    Constant [ int SCARD_E_NO_MEMORY ] { 2148532230 }
    Constant [ int SCARD_E_NO_READERS_AVAILABLE ] { 2148532270 }
    Constant [ int SCARD_E_NO_SERVICE ] { 2148532253 }
    Constant [ int SCARD_E_NO_SMARTCARD ] { 2148532236 }
    Constant [ int SCARD_E_NO_SUCH_CERTIFICATE ] { 2148532268 }
    Constant [ int SCARD_E_NOT_READY ] { 2148532240 }
    Constant [ int SCARD_E_NOT_TRANSACTED ] { 2148532246 }
    Constant [ int SCARD_E_PCI_TOO_SMALL ] { 2148532249 }
    Constant [ int SCARD_E_PROTO_MISMATCH ] { 2148532239 }
    Constant [ int SCARD_E_READER_UNAVAILABLE ] { 2148532247 }
    Constant [ int SCARD_E_READER_UNSUPPORTED ] { 2148532250 }
    Constant [ int SCARD_E_SERVER_TOO_BUSY ] { 2148532273 }
    Constant [ int SCARD_E_SERVICE_STOPPED ] { 2148532254 }
    Constant [ int SCARD_E_SHARING_VIOLATION ] { 2148532235 }
    Constant [ int SCARD_E_SYSTEM_CANCELLED ] { 2148532242 }
    Constant [ int SCARD_E_TIMEOUT ] { 2148532234 }
    Constant [ int SCARD_E_UNEXPECTED ] { 2148532255 }
    Constant [ int SCARD_E_UNKNOWN_CARD ] { 2148532237 }
    Constant [ int SCARD_E_UNKNOWN_READER ] { 2148532233 }
    Constant [ int SCARD_E_UNKNOWN_RES_MNG ] { 2148532267 }
    Constant [ int SCARD_E_UNSUPPORTED_FEATURE ] { 2148532255 }
    Constant [ int SCARD_E_WRITE_TOO_MANY ] { 2148532264 }
    Constant [ int SCARD_F_COMM_ERROR ] { 2148532243 }
    Constant [ int SCARD_F_INTERNAL_ERROR ] { 2148532225 }
    Constant [ int SCARD_F_UNKNOWN_ERROR ] { 2148532244 }
    Constant [ int SCARD_F_WAITED_TOO_LONG ] { 2148532231 }
    Constant [ int SCARD_P_SHUTDOWN ] { 2148532248 }
    Constant [ int SCARD_S_SUCCESS ] { 0 }
    Constant [ int SCARD_W_CANCELLED_BY_USER ] { 2148532334 }
    Constant [ int SCARD_W_CARD_NOT_AUTHENTICATED ] { 2148532335 }
    Constant [ int SCARD_W_CHV_BLOCKED ] { 2148532332 }
    Constant [ int SCARD_W_EOF ] { 2148532333 }
    Constant [ int SCARD_W_REMOVED_CARD ] { 2148532329 }
    Constant [ int SCARD_W_RESET_CARD ] { 2148532328 }
    Constant [ int SCARD_W_SECURITY_VIOLATION ] { 2148532330 }
    Constant [ int SCARD_W_UNPOWERED_CARD ] { 2148532327 }
    Constant [ int SCARD_W_UNRESPONSIVE_CARD ] { 2148532326 }
    Constant [ int SCARD_W_UNSUPPORTED_CARD ] { 2148532325 }
    Constant [ int SCARD_W_WRONG_CHV ] { 2148532331 }
    Constant [ int SCARD_PROTOCOL_RAW ] { 4 }
    Constant [ int SCARD_PROTOCOL_T0 ] { 1 }
    Constant [ int SCARD_PROTOCOL_T1 ] { 2 }
    Constant [ int SCARD_EJECT_CARD ] { 3 }
    Constant [ int SCARD_LEAVE_CARD ] { 0 }
    Constant [ int SCARD_RESET_CARD ] { 1 }
    Constant [ int SCARD_UNPOWER_CARD ] { 2 }
    Constant [ int SCARD_ABSENT ] { 2 }
    Constant [ int SCARD_NEGOTIABLE ] { 32 }
    Constant [ int SCARD_POWERED ] { 16 }
    Constant [ int SCARD_PRESENT ] { 4 }
    Constant [ int SCARD_SPECIFIC ] { 64 }
    Constant [ int SCARD_SWALLOWED ] { 8 }
  }

  - Functions {
    Function [ <internal:PC/SC> function scard_establish_context ] {

      - Parameters [0] {
      }
    }
    Function [ <internal:PC/SC> function scard_release_context ] {

      - Parameters [1] {
        Parameter #0 [ <required> $context ]
      }
    }
    Function [ <internal:PC/SC> function scard_is_valid_context ] {

      - Parameters [1] {
        Parameter #0 [ <required> $context ]
      }
    }
    Function [ <internal:PC/SC> function scard_list_readers ] {

      - Parameters [1] {
        Parameter #0 [ <required> $context ]
      }
    }
    Function [ <internal:PC/SC> function scard_connect ] {

      - Parameters [4] {
        Parameter #0 [ <required> $context ]
        Parameter #1 [ <required> $reader_name ]
        Parameter #2 [ <optional> $preferred_protocol ]
        Parameter #3 [ <optional> &$active_protocol ]
      }
    }
    Function [ <internal:PC/SC> function scard_disconnect ] {

      - Parameters [2] {
        Parameter #0 [ <required> $card ]
        Parameter #1 [ <optional> $disposition ]
      }
    }
    Function [ <internal:PC/SC> function scard_transmit ] {

      - Parameters [2] {
        Parameter #0 [ <required> $card ]
        Parameter #1 [ <required> $command ]
      }
    }
    Function [ <internal:PC/SC> function scard_status ] {

      - Parameters [1] {
        Parameter #0 [ <required> $card ]
      }
    }
    Function [ <internal:PC/SC> function scard_last_errno ] {

      - Parameters [0] {
      }
    }
    Function [ <internal:PC/SC> function scard_errstr ] {

      - Parameters [1] {
        Parameter #0 [ <required> $errno ]
      }
    }
  }
}

```

**Comment by remicollet** [comment]
Notice: this is a minimal fix.

Most "resource" have been moved to opaque object in PHP 8 with type hinting for all functions, and various warning promoted to exception

File: pcsc.c [hunk 1]
```diff
@@ -669,7 +696,7 @@ PHP_FUNCTION(scard_transmit)
   SCARD_IO_REQUEST *recvPci = NULL;
   BYTE *sendBuffer;
   BYTE *recvBuffer;
-  DWORD sendLen, recvLen;
+  DWORD sendLen = 0, recvLen;
```
Comment (remicollet) [inline pcsc.c:699]: but probably the empty string case should be handled

File: pcsc.c [hunk 2]
```diff
@@ -502,6 +525,10 @@ PHP_FUNCTION(scard_establish_context)
   SCARDCONTEXT scard_context = 0;
   LONG rc = 0;
 
+  if (zend_parse_parameters_none() == FAILURE) {
```
Comment (remicollet) [inline pcsc.c:528]: common practice

File: pcsc.c [hunk 3]
```diff
@@ -1,8 +1,6 @@
 /*
  +----------------------------------------------------------------------+
- | PHP Version 5                                                        |
- +----------------------------------------------------------------------+
- | Copyright (c) 1997-2006 The PHP Group                                |
+ | Copyright (c) The PHP Group                                          |
```
Comment (remicollet) [inline pcsc.c:3]: As in main php-src (better to removed than trying to maintain this outdated information)

**Comment by Jan-E** [comment]
Confirmed tot be building for PHP 7.0 up until PHP 8.0.2 on Windows

**Review by NattyNarwhal — APPROVED** [review]
Yeah, this is the change I would have made. (I only needed to test 7.4 at the time.) I'd have probably used the arginfo stubs, LGTM

**Pull request merged**